node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1823 @@
+	ctx: &mut Context,
+	relay_parent: Hash,
+) -> Result<Vec<ValidatorId>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	let query_validators = AllMessages::RuntimeApi(RuntimeApiMessage::Request(
+		relay_parent.clone(),
+		RuntimeApiRequest::Validators(tx),
+	));
+
+	ctx.send_message(query_validators)
+		.await
+		.map_err::<Error, _>(Into::into)?;
+	rx.await
+		.map_err::<Error, _>(Into::into)?
+		.map_err::<Error, _>(Into::into)
+}
+
+/// Query the hash of the `K` ancestors
+async fn query_k_ancestors<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+	k: usize,
+) -> Result<Vec<Hash>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	let query_ancestors = AllMessages::ChainApi(ChainApiMessage::Ancestors {
+		hash: relay_parent,
+		k,
+		response_channel: tx,
+	});
+
+	ctx.send_message(query_ancestors)
+		.await
+		.map_err::<Error, _>(Into::into)?;
+	rx.await
+		.map_err::<Error, _>(Into::into)?
+		.map_err::<Error, _>(Into::into)
+}
+
+/// Query the session index of a relay parent
+async fn query_session_index_for_child<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+) -> Result<SessionIndex>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	let query_session_idx_for_child = AllMessages::RuntimeApi(RuntimeApiMessage::Request(
+		relay_parent.clone(),
+		RuntimeApiRequest::SessionIndexForChild(tx),
+	));
+
+	ctx.send_message(query_session_idx_for_child)
+		.await
+		.map_err::<Error, _>(Into::into)?;
+	rx.await
+		.map_err::<Error, _>(Into::into)?
+		.map_err::<Error, _>(Into::into)
+}
+
+/// Queries up to k ancestors with the constraints of equiv session
+async fn query_up_to_k_ancestors_in_same_session<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+	k: usize,
+) -> Result<Vec<Hash>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	// k + 1 since we always query the child's session index
+	// ordering is [parent, grandparent, greatgrandparent, greatgreatgrandparent, ...]
+	let ancestors = query_k_ancestors(ctx, relay_parent, k + 1).await?;
+	let desired_session = query_session_index_for_child(ctx, relay_parent).await?;
+	// we would only need `ancestors.len() - 1`, but the one extra could avoid a re-alloc
+	// if the consumer wants to push the `relay_parent` onto it too and does not hurt otherwise
+	let mut acc = Vec::with_capacity(ancestors.len());
+
+	// iterate from oldest to youngest
+	let mut iter = ancestors.into_iter().rev();
+	if let Some(oldest) = iter.next() {
+		let mut child_session_index = query_session_index_for_child(ctx, oldest).await?;
+		for ancestor in iter {
+			let ancestor_session_index = child_session_index;
+			child_session_index = query_session_index_for_child(ctx, ancestor).await?;
+
+			if desired_session == ancestor_session_index {
```

Comment:
just to make sure I understand this correctly: once this condition is false, it's always false, right?

Reply:
since we iterator starting from the farthest descendant from `relay_parent`, it will be `false`, until it will be eventually `true` and then stray `true` up to the `relay_parent` (which is not included).